File: packages/sardis-sdk-python/src/sardis_sdk/integrations/langchain.py
"""
LangChain integration for Sardis SDK.

DEPRECATED: This module is a compatibility shim. Use sardis-langchain directly:
    pip install sardis-langchain

    from sardis_langchain import SardisToolkit, SardisPayTool
"""
from __future__ import annotations

import os
import warnings
from decimal import Decimal
from typing import Any, Optional

from pydantic import BaseModel, Field

warnings.warn(
    "sardis_sdk.integrations.langchain is deprecated. "
    "Use sardis-langchain package directly: pip install sardis-langchain\n"
    "  from sardis_langchain import SardisToolkit",
    DeprecationWarning,
    stacklevel=2,
)

try:
    from sardis_langchain import SardisToolkit  # type: ignore
except ImportError:  # pragma: no cover - optional dependency bridge
    SardisToolkit = None  # type: ignore[assignment]


class PayInput(BaseModel):
    """Legacy payment input schema."""

    amount: float = Field(description="Payment amount")
    merchant: str = Field(description="Merchant identifier")
    merchant_address: Optional[str] = Field(default=None, description="Recipient wallet address")
    purpose: str = Field(default="Service payment", description="Reason for payment")
    token: str = Field(default="USDC", description="Token symbol")


class PolicyCheckInput(BaseModel):
    """Legacy policy-check input schema."""

    merchant: str = Field(description="Merchant identifier")
    amount: float = Field(description="Amount to validate")
    purpose: Optional[str] = Field(default=None, description="Optional purpose")


class BalanceCheckInput(BaseModel):
    """Legacy balance-check input schema."""

    token: str = Field(default="USDC", description="Token symbol")
    chain: str = Field(default="base_sepolia", description="Chain identifier")


async def _maybe_await(value: Any) -> Any:
    if hasattr(value, "__await__"):
        return await value
    return value


class _BaseCompatibilityTool:
    name: str = ""
    description: str = ""
    args_schema: type[BaseModel] = BaseModel

    def __init__(
        self,
        *,
        client: Any = None,
        wallet_id: str | None = None,
        agent_id: str | None = None,
        chain: str = "base_sepolia",
        **_: Any,
    ) -> None:
        self.client = client
        self.wallet_id = wallet_id or os.getenv("SARDIS_WALLET_ID", "")
        self.agent_id = agent_id or os.getenv("SARDIS_AGENT_ID", "")
        self.chain = chain


class SardisTool(_BaseCompatibilityTool):
    """Legacy payment tool compatibility class."""

    name = "sardis_pay"
    description = "Execute policy-enforced payment from Sardis wallet."
    args_schema = PayInput

    async def _arun(
        self,
        amount: float,
        merchant: str,
        merchant_address: str | None = None,
        purpose: str = "Service payment",
        token: str = "USDC",
    ) -> str:
        if self.client is None:
            return "Error: Sardis client not initialized"
        if not self.wallet_id:
            return "Error: wallet ID is required"
        if amount <= 0:
            return "Error: amount must be positive"

        destination = merchant_address or merchant
        try:
            result = await _maybe_await(
                self.client.wallets.transfer(
                    self.wallet_id,
                    destination=destination,
                    amount=Decimal(str(amount)),
                    token=token,
                    chain=self.chain,
                    memo=purpose,
                )
            )
            tx_hash = getattr(result, "tx_hash", "")
            chain = getattr(result, "chain", self.chain)
            return f"APPROVED: Payment ${amount:g} to {merchant} submitted on {chain} (tx: {tx_hash})"
        except Exception as exc:
            message = str(exc)
            if any(marker in message.lower() for marker in ("blocked", "policy", "limit", "deny")):
                return f"BLOCKED: PREVENTED by policy - {message}"
            return f"Error: {message}"


class SardisPolicyCheckTool(_BaseCompatibilityTool):
    """Legacy policy-check tool compatibility class."""

    name = "sardis_check_policy"
    description = "Check whether a payment would pass policy."
    args_schema = PolicyCheckInput

    async def _arun(self, merchant: str, amount: float, purpose: str | None = None) -> str:
        if self.client is None:
            return "Error: Sardis client not initialized"
        if not self.wallet_id:
            return "Error: wallet ID is required"

        try:
            wallet = await _maybe_await(self.client.wallets.get(self.wallet_id))
            limit_raw = getattr(wallet, "limit_per_tx", None)
            active = bool(getattr(wallet, "is_active", True))
            limit = Decimal(str(limit_raw)) if limit_raw not in (None, "") else None
            amount_d = Decimal(str(amount))

            if not active:
                return f"WOULD BE BLOCKED (FAIL): wallet disabled for {merchant}"
            if limit is not None and amount_d > limit:
                return (
                    f"WOULD BE BLOCKED (FAIL): amount ${amount:g} exceeds per-tx limit "
                    f"${limit}"
                )
            return f"WOULD BE ALLOWED (PASS): ${amount:g} to {merchant}"
        except Exception as exc:
            return f"Error: {exc}"


class SardisBalanceCheckTool(_BaseCompatibilityTool):
    """Legacy balance tool compatibility class."""

    name = "sardis_check_balance"
    description = "Check current wallet balance."
    args_schema = BalanceCheckInput

    async def _arun(self, token: str = "USDC", chain: str = "base_sepolia") -> str:
        if self.client is None:
            return "Error: Sardis client not initialized"
        if not self.wallet_id:
            return "Error: wallet ID is required"

        try:
            balance = await _maybe_await(
                self.client.wallets.get_balance(self.wallet_id, token=token, chain=chain)
            )
            amount = getattr(balance, "balance", "0")
            return (
                "Wallet Balance\n"
                f"wallet={getattr(balance, 'wallet_id', self.wallet_id)}\n"
                f"token={getattr(balance, 'token', token)}\n"
                f"chain={getattr(balance, 'chain', chain)}\n"
                f"balance={amount}"
            )
        except Exception as exc:
            return f"Error: {exc}"


# Legacy aliases expected by historical integrations/tests
SardisPayTool = SardisTool
SardisCheckPolicyTool = SardisPolicyCheckTool
SardisCheckBalanceTool = SardisBalanceCheckTool


def create_sardis_tools(
    client: Any,
    wallet_id: str,
    *,
    agent_id: str | None = None,
    chain: str = "base_sepolia",
) -> list[_BaseCompatibilityTool]:
    """Legacy tool factory returning payment/policy/balance tools only."""
    common = {"client": client, "wallet_id": wallet_id, "agent_id": agent_id, "chain": chain}
    return [
        SardisTool(**common),
        SardisPolicyCheckTool(**common),
        SardisBalanceCheckTool(**common),
    ]


__all__ = [
    "BalanceCheckInput",
    "PayInput",
    "PolicyCheckInput",
    "SardisBalanceCheckTool",
    "SardisCheckBalanceTool",
    "SardisCheckPolicyTool",
    "SardisPayTool",
    "SardisPolicyCheckTool",
    "SardisTool",
    "SardisToolkit",
    "create_sardis_tools",
]

File: packages/sardis-sdk-js/src/integrations/langchain.ts
/**
 * LangChain.js Integration for Sardis SDK
 *
 * Provides LangChain-compatible tools for AI agents to execute payments
 * through Sardis MPC wallets with policy enforcement.
 *
 * @example
 * ```typescript
 * import { ChatOpenAI } from '@langchain/openai';
 * import { AgentExecutor, createOpenAIFunctionsAgent } from 'langchain/agents';
 * import { SardisClient } from '@sardis/sdk';
 * import { createSardisLangChainTools } from '@sardis/sdk/integrations/langchain';
 *
 * const client = new SardisClient({ apiKey: 'your-api-key' });
 * const tools = createSardisLangChainTools(client, { walletId: 'wallet_123' });
 *
 * const agent = await createOpenAIFunctionsAgent({
 *   llm: new ChatOpenAI(),
 *   tools,
 *   prompt: yourPrompt
 * });
 *
 * const executor = new AgentExecutor({ agent, tools });
 * const result = await executor.invoke({ input: 'Pay $50 to OpenAI' });
 * ```
 */

import { SardisClient } from '../client.js';
import type { ExecutePaymentResponse, WalletBalance, Wallet } from '../types.js';

/**
 * Options for creating Sardis LangChain tools
 */
export interface SardisLangChainOptions {
    /** Default wallet ID to use for payments */
    walletId?: string;
    /** Default agent ID */
    agentId?: string;
    /** Default chain for transactions */
    chain?: 'base' | 'base_sepolia' | 'polygon' | 'polygon_amoy' | 'ethereum' | 'ethereum_sepolia';
    /** Default token */
    token?: 'USDC' | 'USDT' | 'PYUSD' | 'EURC';
}

/**
 * LangChain tool interface
 */
interface LangChainTool {
    name: string;
    description: string;
    schema: {
        type: 'object';
        properties: Record<string, unknown>;
        required?: string[];
    };
    func: (input: Record<string, unknown>) => Promise<string>;
}

/**
 * Generate a unique mandate ID
 */
function generateMandateId(): string {
    const timestamp = Date.now().toString(36);
    const random = Math.random().toString(36).substring(2, 10);
    return `mnd_${timestamp}${random}`;
}

/**
 * Create a SHA-256 hash for audit purposes
 */
async function createAuditHash(data: string): Promise<string> {
    if (typeof crypto !== 'undefined' && crypto.subtle) {
        const encoder = new TextEncoder();
        const dataBuffer = encoder.encode(data);
        const hashBuffer = await crypto.subtle.digest('SHA-256', dataBuffer);
        const hashArray = Array.from(new Uint8Array(hashBuffer));
        return hashArray.map(b => b.toString(16).padStart(2, '0')).join('');
    }
    return `hash_${Date.now().toString(16)}`;
}

/**
 * Create LangChain-compatible tools for Sardis payments.
 *
 * @param client - Initialized SardisClient instance
 * @param options - Configuration options
 * @returns Array of LangChain-compatible tools
 */
export function createSardisLangChainTools(
    client: SardisClient,
    options: SardisLangChainOptions = {}
): LangChainTool[] {
    const defaultWalletId = options.walletId || process.env.SARDIS_WALLET_ID || '';
    const defaultAgentId = options.agentId || process.env.SARDIS_AGENT_ID || '';
    const defaultChain = options.chain || 'base_sepolia';
    const defaultToken = options.token || 'USDC';

    const tools: LangChainTool[] = [
        {
            name: 'sardis_pay',
            description: 'Execute a secure payment using Sardis MPC wallet. Validates against spending policies. Use this when the user wants to pay a vendor or service.',
            schema: {
                type: 'object',
                properties: {
                    amount: {
                        type: 'number',
                        description: 'The amount to pay in USD (or token units)'
                    },
                    vendor: {
                        type: 'string',
                        description: 'The name of the merchant or service provider'
                    },
                    vendor_address: {
                        type: 'string',
                        description: 'The wallet address of the vendor (0x...). Optional - will resolve if not provided.'
                    },
                    purpose: {
                        type: 'string',
                        description: 'The reason for the payment, used for policy validation'
                    },
                    token: {
                        type: 'string',
                        enum: ['USDC', 'USDT', 'PYUSD', 'EURC'],
                        description: 'The stablecoin to use. Defaults to USDC.'
                    }
                },
                required: ['amount', 'vendor']
            },
            func: async (input: Record<string, unknown>): Promise<string> => {
                const amount = input.amount as number;
                const vendor = input.vendor as string;
                const vendorAddress = input.vendor_address as string | undefined;
                const purpose = input.purpose as string | undefined;
                const token = (input.token as 'USDC' | 'USDT' | 'PYUSD' | 'EURC') || defaultToken;

                if (!defaultWalletId) {
                    return JSON.stringify({
                        success: false,
                        error: 'No wallet ID configured. Set walletId in options or SARDIS_WALLET_ID env var.'
                    });
                }

                try {
                    const mandateId = generateMandateId();
                    const timestamp = new Date().toISOString();
                    const amountMinor = Math.round(amount * 1_000_000).toString();

                    const auditData = JSON.stringify({
                        mandate_id: mandateId,
                        subject: defaultWalletId,
                        destination: vendorAddress || `pending:${vendor}`,
                        amount_minor: amountMinor,
                        token,
                        purpose: purpose || `Payment to ${vendor}`,
                        timestamp
                    });
                    const auditHash = await createAuditHash(auditData);

                    const mandate = {
                        mandate_id: mandateId,
                        subject: defaultWalletId,
                        destination: vendorAddress || `pending:${vendor}`,
                        amount_minor: amountMinor,
                        token,
                        chain: defaultChain,
                        purpose: purpose || `Payment to ${vendor}`,
                        vendor_name: vendor,
                        agent_id: defaultAgentId,
                        timestamp,
                        audit_hash: auditHash,
                        metadata: {
                            vendor,
                            category: 'saas',
                            initiated_by: 'ai_agent',
                            tool: 'langchain_js'
                        }
                    };

                    const result: ExecutePaymentResponse = await client.payments.executeMandate(mandate);

                    return JSON.stringify({
                        success: true,
                        status: result.status,
                        payment_id: result.payment_id,
                        transaction_hash: result.tx_hash,
                        chain: result.chain,
                        ledger_tx_id: result.ledger_tx_id,
                        message: `Payment of $${amount} ${token} to ${vendor} ${result.status === 'completed' ? 'completed' : 'initiated'}.`
                    });
                } catch (error: unknown) {
                    const errorMessage = error instanceof Error ? error.message : 'Payment failed';

                    if (errorMessage.includes('policy') || errorMessage.includes('blocked') || errorMessage.includes('limit')) {
                        return JSON.stringify({
                            success: false,
                            blocked: true,
                            error: errorMessage,
                            message: `Payment to ${vendor} blocked by policy: ${errorMessage}`
                        });
                    }

                    return JSON.stringify({
                        success: false,
                        error: errorMessage
                    });
                }
            }
        },

        {
            name: 'sardis_check_balance',
            description: 'Check the current balance of a Sardis wallet. Use this before making payments to ensure sufficient funds.',
            schema: {
                type: 'object',
                properties: {
                    token: {
                        type: 'string',
                        enum: ['USDC', 'USDT', 'PYUSD', 'EURC'],
                        description: 'The token to check balance for. Defaults to USDC.'
                    },
                    chain: {
                        type: 'string',
                        description: 'The blockchain to check balance on. Defaults to base_sepolia.'
                    }
                }
            },
            func: async (input: Record<string, unknown>): Promise<string> => {
                const token = (input.token as 'USDC' | 'USDT' | 'PYUSD' | 'EURC') || defaultToken;
                const chain = (input.chain as string) || defaultChain;

                if (!defaultWalletId) {
                    return JSON.stringify({
                        success: false,
                        error: 'No wallet ID configured'
                    });
                }

                try {
                    const balance: WalletBalance = await client.wallets.getBalance(
                        defaultWalletId,
                        chain,
                        token
                    );

                    return JSON.stringify({
                        success: true,
                        wallet_id: balance.wallet_id,
                        balance: balance.balance,
                        token: balance.token,
                        chain: balance.chain,
                        address: balance.address
                    });
                } catch (error: unknown) {
                    return JSON.stringify({
                        success: false,
                        error: error instanceof Error ? error.message : 'Failed to get balance'
                    });
                }
            }
        },

        {
            name: 'sardis_get_wallet',
            description: 'Get information about the Sardis wallet including spending limits and policy settings.',
            schema: {
                type: 'object',
                properties: {}
            },
            func: async (): Promise<string> => {
                if (!defaultWalletId) {
                    return JSON.stringify({
                        success: false,
                        error: 'No wallet ID configured'
                    });
                }

                try {
                    const wallet: Wallet = await client.wallets.get(defaultWalletId);

                    return JSON.stringify({
                        success: true,
                        wallet: {
                            id: wallet.id,
                            agent_id: wallet.agent_id,
                            currency: wallet.currency,
                            limit_per_tx: wallet.limit_per_tx,
                            limit_total: wallet.limit_total,
                            is_active: wallet.is_active,
                            addresses: wallet.addresses
                        }
                    });
                } catch (error: unknown) {
                    return JSON.stringify({
                        success: false,
                        error: error instanceof Error ? error.message : 'Failed to get wallet'
                    });
                }
            }
        },

        {
            name: 'sardis_check_policy',
            description: 'Check if a payment would be allowed by the spending policy without executing it. Use this to validate payments before execution.',
            schema: {
                type: 'object',
                properties: {
                    amount: {
                        type: 'number',
                        description: 'The amount to pay in USD'
                    },
                    vendor: {
                        type: 'string',
                        description: 'The name of the merchant or service provider'
                    },
                    purpose: {
                        type: 'string',
                        description: 'The reason for the payment'
                    }
                },
                required: ['amount', 'vendor']
            },
            func: async (input: Record<string, unknown>): Promise<string> => {
                const amount = input.amount as number;
                const vendor = input.vendor as string;
                const purpose = input.purpose as string | undefined;

                if (!defaultWalletId) {
                    return JSON.stringify({
                        success: false,
                        error: 'No wallet ID configured'
                    });
                }

                try {
                    // Get wallet to check limits
                    const wallet: Wallet = await client.wallets.get(defaultWalletId);
                    const limitPerTx = parseFloat(wallet.limit_per_tx);
                    const limitTotal = parseFloat(wallet.limit_total);

                    // Basic policy checks
                    const checks: { name: string; passed: boolean; reason?: string }[] = [];

                    // Per-transaction limit check
                    if (amount <= limitPerTx) {
                        checks.push({ name: 'per_transaction_limit', passed: true });
                    } else {
                        checks.push({
                            name: 'per_transaction_limit',
                            passed: false,
                            reason: `Amount $${amount} exceeds per-transaction limit of $${limitPerTx}`
                        });
                    }

                    // Wallet active check
                    if (wallet.is_active) {
                        checks.push({ name: 'wallet_active', passed: true });
                    } else {
                        checks.push({
                            name: 'wallet_active',
                            passed: false,
                            reason: 'Wallet is not active'
                        });
                    }

                    const allPassed = checks.every(c => c.passed);

                    return JSON.stringify({
                        success: true,
                        allowed: allPassed,
                        checks,
                        summary: allPassed
                            ? `Payment of $${amount} to ${vendor} would be allowed`
                            : `Payment of $${amount} to ${vendor} would be blocked: ${checks.filter(c => !c.passed).map(c => c.reason).join('; ')}`
                    });
                } catch (error: unknown) {
                    return JSON.stringify({
                        success: false,
                        error: error instanceof Error ? error.message : 'Failed to check policy'
                    });
                }
            }
        }
    ];

    return tools;
}

/**
 * Create a single LangChain tool for payments only.
 * Use this if you only need the payment functionality.
 */
export function createSardisPaymentTool(
    client: SardisClient,
    options: SardisLangChainOptions = {}
): LangChainTool {
    const tools = createSardisLangChainTools(client, options);
    return tools.find(t => t.name === 'sardis_pay')!;
}

export type { LangChainTool };
